### ISSUE TITLE
[RabbitMQ Microservice] "PRECONDITION_FAILED - reply consumer already set"

### ISSUE THREAD

**Issue by darioxtx** [nestjs/nest#3138]
## Current behavior

Sometimes I'm getting these errors and after that microcervice client stops working and no way to catch these errors and I have to restart application to work properly again.

Issue is similar to [#2323](https://github.com/nestjs/nest/issues/2323) and [#2337](https://github.com/nestjs/nest/issues/2337)

```ts
AssertionError [ERR_ASSERTION]: "0 == 1"
    at Mux._readIncoming (app\\node_modules\\amqplib\\lib\\mux.js:92:12)
    at Immediate._onImmediate (app\\node_modules\\amqplib\\lib\\mux.js:112:12)
    at runCallback (timers.js:810:20)
    at tryOnImmediate (timers.js:768:5)
    at processImmediate [as _immediateCallback] (timers.js:745:5)
	
Error: Operation failed: BasicConsume; 406 (PRECONDITION-FAILED) with message "PRECONDITION_FAILED - reply consumer already set"
    at reply (app\\node_modules\\amqplib\\lib\\channel.js:134:29)
    at ConfirmChannel.C.accept (app\\node_modules\\amqplib\\lib\\channel.js:417:7)
	at Connection.mainAccept [as accept] (app\\node_modules\\amqplib\\lib\\connection.js:64:33)
    at Socket.go (app\\node_modules\\amqplib\\lib\\connection.js:478:48)
    at emitNone (events.js:106:13)
    at Socket.emit (events.js:208:7)
    at emitReadable_ (_stream_readable.js:513:10)
    at emitReadable (_stream_readable.js:507:7)
    at addChunk (_stream_readable.js:274:7)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:601:20)
	From previous event:
    at ConfirmChannel.C.consume (app\\node_modules\\amqplib\\lib\\channel_model.js:170:18)
    at ChannelWrapper.channel.addSetup (app\\node_modules\\@nestjs\\microservices\\client\\client-rmq.js:37:52)
    at app\\node_modules\\promise-breaker\\index.js:265:46
    at <anonymous>
    at process._tickCallback (internal/process/next_tick.js:189:7)
```

## Input Code

To get this error every time, you need to wait some time while processing message for ex.: long running object serializer.

```ts
@Module({
  imports: [
    ClientsModule.register([
      {
        name: 'RMQ_SERVICE',
        transport: Transport.RMQ,
        options: {
          urls: [`amqp://localhost:5672`],
          queue: "log_queue",
          queueOptions: { durable: true },
          serializer: {
            serialize: (value) => {
              const dt = new Date();
              while (new Date().getTime() - dt.getTime() <= 15000) { /* Do nothing */ }

              return value;
            },
          },
        },
      },
    ]),
  ],
  controllers: [AppController],
  providers: [AppService],
  exports: [AppService],
})
export class AppModule implements NestModule {
  configure(consumer: MiddlewareConsumer) {    
  }
}

@Injectable()
export class AppService {
  constructor(@Inject('RMQ_SERVICE') private readonly client: ClientProxy ) {  }

  async getRMQServiceStatus(): Promise<string> {
    const pattern = { cmd: 'status' };
    const payload = 1;
    return this.client.send<string>(pattern, payload).toPromise();
  }

  async logAction(model: SomeModel): Promise<void> {
    const pattern ='log_action'
    await this.client.emit<void>(pattern, model);
  }
}
```

## Expected behavior

Client should catch this error and throw to application and should reconnect and start work again.

## Environment

<pre><code>
Nest version: 6.8.2

 
For Tooling issues:
- Node version: v8.16.0  
- Platform:  Linux, Windows

Others:

</code></pre>

**Comment by kamilmysliwiec** [comment]
Please, provide a minimal repository which reproduces your issue.

**Comment by darioxtx** [comment]
Hi @kamilmysliwiec,

[Here](https://github.com/darioxtx/nestjs-rmq-issue) you can find minimal repository which reproduces this issue.

```bash
npm install
npm run build
npm run start:prod:server
npm run start:prod
```

To test if microservice works fine use http://127.0.0.1:3030/status;
If you get response then try to run http://127.0.0.1:3030/test (run few times if you do not get error, or increase wait time).
After the error client will stop working.

**Comment by enissay-git** [comment]
Hello @darioxtx,

I'am looking at your issue :) (first time trying to contribute)

It would have been nice to have a docker-compose.yml for testing locally on your project.

I will do it on my side locally but this could improve your issue and your example project.

**Comment by darioxtx** [comment]
Hi @enissay-git,

Sorry but I do not work with docker-compose, so it will be hard to make such a file

**Comment by enissay-git** [comment]
Hi @darioxtx if i made something working with docker-compose i will keep you in touch :)

**Comment by MichaelKoch** [comment]
ans news about that issue ?
Facing the same Problem..

**Comment by enissay-git** [comment]
Hello @MichaelKoch not from my side, i'am trying to launch the example project issue but there is two command to launch so i need some clarification about the difference between "npm run start:prod:server" and "npm run start:prod" from @darioxtx

**Comment by MichaelKoch** [comment]
Hi @enissay-git,
maybe ... we can collaborate the bring up the sample for the root cause ....
Give me some time to do some discussion ... and get some decissions (inhouse) 
In the mean time : Whats the issue to run the example ? 

br 
Michael

**Comment by darioxtx** [comment]
Hi @enissay-git,

Sample project has 2 "microservices". First microservice use RMQ transport and second is as a client which makes a requests to first microservice. if you run all microservices as one command using `await app.startAllMicroservicesAsync();` you won't be able to reproduce this issue. But in real world you run microservices as a separate processes ex.: kubernetes. 

In this example you have to debug client microservice `npm run start:prod`.

**Comment by enissay-git** [comment]
It would be nice to collaborate yes :), @MichaelKoch  the issue is that i try to run it with a docker-compose to make it easy to reproduce and have a rabbitmq server + the example issue projet running together but due to the two command to run i have some trouble to make it work.

**Comment by enissay-git** [comment]
Thanks for the details @darioxtx i will try with that info :)

**Comment by darioxtx** [comment]
Not sure why you complicating things with docker-compose, but if you want so, make 2 docker images, one should run `npm run start:prod:server` an other `npm run start:prod` :)

**Comment by darioxtx** [comment]
Hi guys, any progress on this issue?

**Comment by enissay-git** [comment]
Hi @darioxtx not yet from my side sorry (i don't found time to work on it for now on my free time)

**Comment by sergey-telpuk** [comment]
@kamilmysliwiec @darioxtx @enissay-git @paolomainardi 
Hello guys, look at [this pull request](https://github.com/nestjs/nest/pull/3581)

**Comment by darioxtx** [comment]
@sergey-telpuk @kamilmysliwiec, thank you guys

**Comment by kamilmysliwiec** [comment]
Fixed in 6.10.9. Thanks @sergey-telpuk!
